### PR TITLE
[WIP] InputCommon

### PR DESCRIPTION
**Pull request by shonumi**
The purpose of InputCommon is to centralize input handling from a single interface. It also abstracts the process to allow different backends (SDL, DInput, XInput) to write their own implementations. Currently only SDL has an implementation written here, but others can be added.

Please keep in mind that this represents basic work done towards getting InputCommon towards where it ideally needs to be. Emu_window still handles keyboard input for GLFW and Qt (this is something InputCommon should do), and there is no configuration just yet. For those brave enough to test controller support, the default is set to work with an XBOX 360 controller (Pretty much 1:1 A on 360 = A on the 3DS and so on).

### CODE REVIEW

File: src/input_common/sdl_input/sdl_input.h [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    bool Init();
+
+    ///Grabs a list of devices supported by the backend
+    void DiscoverDevices();
+
+    ///Grab and process input via SDL2
+    void Poll();
+
+    ///Shuts down all joysticks opened by SDL2
+    void ShutDown();
```
Comment (lioncash) [inline src/input_common/sdl_input/sdl_input.h]: Init, DiscoverDevice, Poll, and ShutDown should have an override specifier on them.

File: src/input_common/sdl_input/sdl_input.h [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+#pragma once
+
+#include <SDL2/SDL.h>
+#include <iostream>
+#include "input_common/input_common.h"
+
+namespace InputCommon {
+
+class SDLController  : virtual public ControllerBase {
```
Comment (lioncash) [inline src/input_common/sdl_input/sdl_input.h]: `class SDLController final : public ControllerBase`

File: src/input_common/sdl_input/sdl_input.cpp [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+    //Attempt to close joystick with SDL
+    if(jpad != nullptr) {
+        SDL_JoystickClose(jpad);
+    }
+}
+
+void SDLController::Poll() {
+    if(SDL_PollEvent(&input_event)) {
+        if((input_event.type == SDL_JOYBUTTONDOWN) || (input_event.type == SDL_JOYBUTTONUP)
+        || (input_event.type == SDL_JOYAXISMOTION) || (input_event.type == SDL_JOYHATMOTION)) {
```
Comment (lioncash) [inline src/input_common/sdl_input/sdl_input.cpp]: the `==` conditions don't need to be parenthesized.

File: src/input_common/input_common.cpp [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        case NO_JOY:
+            break;
+
+        //What?
+        default:
+            break;
+    }
+
+    //Shutdown immediately if backend failed to initialize
+    if((g_user_input != nullptr) && (!g_user_input->activated)) {
```
Comment (lioncash) [inline src/input_common/input_common.cpp]: The conditions here don't need to be internally parenthesized.

File: src/input_common/input_common.cpp [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+    //Setup CoreTiming
+    frame_ticks = 268123480 / Settings::values.gpu_refresh_rate / 16;
+
+    input_event = CoreTiming::RegisterEvent("InputCommon::InputCallback", InputCallback);
+    CoreTiming::ScheduleEvent(frame_ticks, input_event);
+}
+
+void ShutDown() {
+    if((g_user_input != nullptr)
```
Comment (lioncash) [inline src/input_common/input_common.cpp]: extra parenthesis?

**Comment by lioncash** [comment]
`if` and `switch` should also have a space after them.

File: src/input_common/input_common.cpp (hunk 5 above)
Comment (shonumi) [inline src/input_common/input_common.cpp]: Whoops :p

File: src/input_common/input_common.h [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    std::string device_name;
+
+    ///Internal view this specific object will have of 3DS input states
+    ControllerState controller;
+};
+
+///Initialize the user input system
+void Init(ControllerBackends backend);
+
+///Deactive the user input system
```
Comment (purpasmart96) [inline src/input_common/input_common.h]: Space after the "///"

**Comment by shonumi** [comment]
Sorry about the stylistic mess, I'm still stuck on other styles I've used for other projects :p 

Let me know about anything else that catches your eyes. Thanks!

File: src/input_common/sdl_input/sdl_input.h [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    void ProcessInput();
+
+    /// Sets up keymaps from input configuration
+    void SetupKeyMaps();
+
+    /// Sets analog deadzones
+    void SetDeadZone(int range);
+
+    /// Checks if analog input is within the dead zone
+    bool CheckDeadZone(int range);
```
Comment (lioncash) [inline src/input_common/sdl_input/sdl_input.h]: CheckDeadZone can be const.

File: src/input_common/input_common.h [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    virtual void Poll() = 0;
+
+    /// Shuts down all backend related data
+    virtual void Shutdown() = 0;
+
+    /// Returns internal name of currently selected device. Expose this to the UI
+    std::string GetDeviceName() const;
+
+    /// Returns internal pad state
+    const Service::HID::PadState GetPadState() const;
```
Comment (lioncash) [inline src/input_common/input_common.h]: The first const isn't necessary, since it isn't returning a pointer.

File: src/core/hle/service/hid/hid.cpp [hunk 9]
```diff
@@ -49,7 +51,7 @@ static u32 next_touch_index;
 
 void Update() {
     SharedMem* mem = reinterpret_cast<SharedMem*>(shared_mem->GetPointer());
-    const PadState state = VideoCore::g_emu_window->GetPadState();
+    const PadState state = (InputCommon::g_user_input == nullptr) ? VideoCore::g_emu_window->GetPadState() : InputCommon::g_user_input->GetPadState();
```
Comment (bunnei) [inline src/core/hle/service/hid/hid.cpp]: Line width should be < 100 characters

File: src/core/system.cpp [hunk 10]
```diff
@@ -22,9 +24,11 @@ void Init(EmuWindow* emu_window) {
     Kernel::Init();
     HLE::Init();
     VideoCore::Init(emu_window);
+    InputCommon::Init(InputCommon::SDL2_JOY);
```
Comment (bunnei) [inline src/core/system.cpp]: Input backend should be a configuration setting

File: src/input_common/controller_state.h [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include "common/common_types.h"
+#include "core/hle/service/hid/hid.h"
+
+namespace InputCommon {
+
+    struct ControllerState {
```
Comment (bunnei) [inline src/input_common/controller_state.h]: Don't indent things within a namespace

File: src/input_common/input_common.cpp [hunk 12]
```diff
@@ -0,0 +1,85 @@
+// Copyright 2015 Citra Emulator Project
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
+
+#include "input_common/input_common.h"
+
+#ifdef HAS_SDL
```
Comment (bunnei) [inline src/input_common/input_common.cpp]: It's nice that you made SDL optional, but having multiple dependency paths (long term) will probably be a pain. I think I'd prefer that we nix GLFW (what we use for the CLI window) and replace with SDL2, and make it a mandatory dependency.


Perhaps OK as an incremen

File: src/input_common/sdl_input/sdl_input.cpp [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+void SDLController::Poll() {
+    if (SDL_PollEvent(&input_event)) {
+        if (input_event.type == SDL_JOYBUTTONDOWN || input_event.type == SDL_JOYBUTTONUP
+        || input_event.type == SDL_JOYAXISMOTION || input_event.type == SDL_JOYHATMOTION) {
+            ProcessInput();
+        }
+    }
+}
+
+void SDLController::ProcessInput() {
```
Comment (linkmauve) [inline src/input_common/sdl_input/sdl_input.cpp]: Instead of defining input_event as an object variable, pass it as argument here.

File: src/input_common/input_common.cpp [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+    #include "input_common/sdl_input/sdl_input.h"
+#endif
+
+#include "core/core_timing.h"
+#include "core/settings.h"
+
+namespace InputCommon {
+
+//Set to nullptr until initialized by a backend
```
Comment (bunnei) [inline src/input_common/input_common.cpp]: Space between the `//` and the comment

File: src/input_common/sdl_input/sdl_input.cpp [hunk 15]
```diff
@@ -0,0 +1,238 @@
...
+
+void SDLController::Shutdown() {
+    //Attempt to close joystick with SDL
+    if (jpad != nullptr) {
+        SDL_JoystickClose(jpad);
+    }
+}
+
+void SDLController::Poll() {
+    if (SDL_PollEvent(&input_event)) {
```
Comment (linkmauve) [inline src/input_common/sdl_input/sdl_input.cpp]: s/if/while/, because there could be multiple events in queue.

File: src/input_common/sdl_input/sdl_input.cpp [hunk 16]
```diff
@@ -0,0 +1,238 @@
...
+    KeyMap::SetKeyMapping({202, joypad_id}, Service::HID::PAD_CIRCLE_UP);
+    KeyMap::SetKeyMapping({203, joypad_id}, Service::HID::PAD_CIRCLE_DOWN);
+}
+
+void SDLController::SetDeadZone(int range) {
+    dead_zone = range;
+}
+
+bool SDLController::CheckDeadZone(int range) const {
+    if ((range < -dead_zone) || (range > dead_zone)) {
```
Comment (linkmauve) [inline src/input_common/sdl_input/sdl_input.cpp]: Check against std::abs(dead_zone) instead.

File: src/input_common/sdl_input/sdl_input.cpp [hunk 17]
```diff
@@ -0,0 +1,238 @@
...
+        return true;
+    } else if ((range > -dead_zone) && (range < dead_zone)) {
```
Comment (linkmauve) [inline src/input_common/sdl_input/sdl_input.cpp]: Drop the else as it happens after a return, and in every other case you want to return false.
Also no brackets since the if contains only one line.

File: src/input_common/input_common.cpp [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+    return device_name;
+}
+
+Service::HID::PadState ControllerBase::GetPadState() const {
+    return controller.pad_state;
+}
+
+void Init(ControllerBackends backend) {
+    switch (backend) {
+        #ifdef HAS_SDL
```
Comment (bunnei) [inline src/input_common/input_common.cpp]: Ugly... Seeing this reaffirms to me that this PR should be contingent on SDL2 being a mandatory dependency, and it should probably just replace GLFW. Could you submit such  a PR as a prereq to this one? It should be quite trivial.

File: src/input_common/input_common.cpp [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+            //SDL2 backend selected
+            case SDL2_JOY:
+                g_user_input = new SDLController();
+                break;
+        #endif
+
+        //No backend selected
```
Comment (bunnei) [inline src/input_common/input_common.cpp]: FWIW re: the comment above, please search+replace all your comments to be `// comment` (with the space)

File: src/input_common/input_common.cpp [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+        case NO_JOY:
+            break;
+
+        //What?
```
Comment (bunnei) [inline src/input_common/input_common.cpp]: Bad comment, probably unnecessary. Default case statement should log an error (probably at the "critical" level)

File: src/input_common/input_common.cpp [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+static int input_event;
+
+//Ticks for CoreTiming event, same as VBlank in GPU
+static u64 frame_ticks;
+
+/// Callback for CoreTiming
+static void InputCallback(u64 userdata, int cycles_late) {
+
+    //Call user input plugin Poll()
+    if (g_user_input != nullptr) {
```
Comment (bunnei) [inline src/input_common/input_common.cpp]: nullptr check here is not very elegant... I would initialize `g_user_input` at runtime, then add an `IsValid()` (or something similar) method to do this check.

File: src/input_common/input_common.cpp (hunk 12 above)
Comment (linkmauve) [inline src/input_common/input_common.cpp]: I think this can be done later, there should be no issue with using both at once.

I’m not sure SDL2 should be mandatory, we could just make input_common expose no joystick at all, right?

File: src/input_common/input_common.cpp (hunk 21 above)
Comment (bunnei) [inline src/input_common/input_common.cpp]: ditto with the other checks below...

File: src/input_common/sdl_input/sdl_input.cpp [hunk 22]
```diff
@@ -0,0 +1,238 @@
...
+            controller.pad_state.hex &= ~KeyMap::GetPadKey({pad_id, joypad_id}).hex;
+            break;
+
+        //Joystick axis motion events
+        case SDL_JOYAXISMOTION:
+            pad_id = 200;
+            pad_id += (input_event.jaxis.axis * 2);
+            val = input_event.jaxis.value;
+
+            if (val > 0) {
```
Comment (linkmauve) [inline src/input_common/sdl_input/sdl_input.cpp]: Same here, and in many other places, we prefer not to have unnecessary brackets when there is only one line in the then (*and* else) part(s).

File: src/input_common/sdl_input/sdl_input.cpp [hunk 23]
```diff
@@ -0,0 +1,238 @@
...
+}
+
+void SDLController::DiscoverDevices() {}
+
+void SDLController::Shutdown() {
+    //Attempt to close joystick with SDL
+    if (jpad != nullptr) {
+        SDL_JoystickClose(jpad);
+    }
+}
```
Comment (linkmauve) [inline src/input_common/sdl_input/sdl_input.cpp]: You should also call SDL_Quit() here.

File: src/input_common/sdl_input/sdl_input.cpp [hunk 24]
```diff
@@ -0,0 +1,238 @@
...
+    if (SDL_Init(SDL_INIT_JOYSTICK) != 0) {
+        return false;
+    }
+
+    //Attempt to open joystick with SDL
+    //TODO - Use DiscoverDevice to generate a list of available joysticks, then open the correct index
+    jpad = SDL_JoystickOpen(index);
+
+    if (jpad == nullptr) {
+        return false;
```
Comment (linkmauve) [inline src/input_common/sdl_input/sdl_input.cpp]: Call Shutdown() here, probably.

File: src/input_common/input_common.h [hunk 25]
```diff
@@ -0,0 +1,60 @@
...
+#pragma once
+
+#include "common/common_types.h"
+#include "controller_state.h"
+
+namespace InputCommon {
+
+/// Enum defining available backends
+enum ControllerBackends {
+    NO_JOY,
```
Comment (bunnei) [inline src/input_common/input_common.h]: `JOY` naming is _so_ GameCube scene :D How about making this an `enum class`, then renaming `NO_JOY` to `NONE` and `SDL2_JOY` to `SDL2`? That way, these enumerations will be explicitly referenced (e.g. `ControllerBackend::SDL2`), so the "JOY" part is now redundant.

File: src/input_common/input_common.h [hunk 26]
```diff
@@ -0,0 +1,60 @@
...
+
+#pragma once
+
+#include "common/common_types.h"
+#include "controller_state.h"
+
+namespace InputCommon {
+
+/// Enum defining available backends
+enum ControllerBackends {
```
Comment (bunnei) [inline src/input_common/input_common.h]: `InputBackend` is a more consistent name IMO

File: src/input_common/sdl_input/sdl_input.cpp [hunk 27]
```diff
@@ -0,0 +1,238 @@
...
+    return true;
+}
+
+void SDLController::DiscoverDevices() {}
+
+void SDLController::Shutdown() {
+    //Attempt to close joystick with SDL
+    if (jpad != nullptr) {
+        SDL_JoystickClose(jpad);
+    }
```
Comment (linkmauve) [inline src/input_common/sdl_input/sdl_input.cpp]: Set back jpad to nullptr, so other checks like that won’t take the wrong branch.

File: src/input_common/input_common.h [hunk 28]
```diff
@@ -0,0 +1,60 @@
...
+    NO_JOY,
+    SDL2_JOY
+};
+
+class ControllerBase {
```
Comment (bunnei) [inline src/input_common/input_common.h]: `InputBase` ?

Comment (bunnei) [inline src/input_common/input_common.h]: Also, this class should inherit from `NonCopyable`

File: src/core/hle/service/hid/hid.cpp (hunk 9 above)
Comment (linkmauve) [inline src/core/hle/service/hid/hid.cpp]: Also add a comment saying it’s only a stop-gap solution until both keyboard backends are ported to InputCommon.

File: externals/cmake-modules/FindSDL2.cmake [hunk 29]
```diff
@@ -0,0 +1,180 @@
...
+# see accompanying file Copyright.txt for details.
+#
+# This software is distributed WITHOUT ANY WARRANTY; without even the
+# implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the License for more information.
+#=============================================================================
+# (To distribute this file outside of CMake, substitute the full
+#  License text for the above reference.)
+
+FIND_PATH(SDL2_INCLUDE_DIR SDL.h
```
Comment (linkmauve) [inline externals/cmake-modules/FindSDL2.cmake]: I find this solution very poor, at least on UNIX-like the path to the include directory shouldn’t be hardcoded, you should instead use pkg-config.

File: externals/cmake-modules/FindSDL2.cmake [hunk 30]
```diff
@@ -0,0 +1,180 @@
...
+  /usr/local/include/SDL2
+  /usr/include/SDL2
+  /sw # Fink
+  /opt/local # DarwinPorts
+  /opt/csw # Blastwave
+  /opt
+)
+#MESSAGE("SDL2_INCLUDE_DIR is ${SDL2_INCLUDE_DIR}")
+
+FIND_LIBRARY(SDL2_LIBRARY_TEMP
```
Comment (linkmauve) [inline externals/cmake-modules/FindSDL2.cmake]: Same here.

File: src/input_common/sdl_input/sdl_input.h [hunk 31]
```diff
@@ -0,0 +1,57 @@
+// Copyright 2015 Citra Emulator Project
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include <SDL2/SDL.h>
```
Comment (linkmauve) [inline src/input_common/sdl_input/sdl_input.h]: #include <SDL.h>, so that users who put it elsewhere than in a SDL2 directory can still include it.

File: src/input_common/sdl_input/sdl_input.cpp [hunk 32]
```diff
@@ -0,0 +1,238 @@
...
+                    controller.pad_state.hex &= ~KeyMap::GetPadKey({pad_id+2, joypad_id}).hex;
+                    controller.pad_state.hex |= KeyMap::GetPadKey({pad_id+3, joypad_id}).hex;
+                    break;
+            }
+
+            break;
+    }
+}
+
+void SDLController::SetupKeyMaps() {
```
Comment (linkmauve) [inline src/input_common/sdl_input/sdl_input.cpp]: It would be good to add configurability right now, you just have to use Settings::values.pad_*_key instead of hardcoding them.

Comment (linkmauve) [inline src/input_common/sdl_input/sdl_input.cpp]: Maybe you should use the SDL_GameController API instead of the SDL_Joystick one, and depend on such a database: https://github.com/gabomdq/SDL_GameControllerDB

File: src/input_common/input_common.h [hunk 33]
```diff
@@ -0,0 +1,60 @@
...
+};
+
+/// Initialize the user input system
+void Init(ControllerBackends backend);
+
+/// Deactive the user input system
+void Shutdown();
+
+/// InputCommon 'plugin'
+extern ControllerBase* g_user_input;
```
Comment (neobrain) [inline src/input_common/input_common.h]: use unique_ptr instead

File: src/input_common/input_common.h (hunk 25 above)
Comment (neobrain) [inline src/input_common/input_common.h]: +1.

A more modern terminology would be "gamepad" rather than "joy", btw.

File: src/input_common/input_common.h [hunk 34]
```diff
@@ -0,0 +1,60 @@
...
+    virtual void DiscoverDevices() = 0;
+
+    /// Grab and process input
+    virtual void Poll() = 0;
+
+    /// Shuts down all backend related data
+    virtual void Shutdown() = 0;
+
+    /// Returns internal name of currently selected device. Expose this to the UI
+    std::string GetDeviceName() const;
```
Comment (neobrain) [inline src/input_common/input_common.h]: When returning non-trivial objects from functions, make them const so that people don't accidentally use them as references.

Comment (neobrain) [inline src/input_common/input_common.h]: Actually, this can even return a const reference.

File: src/input_common/sdl_input/sdl_input.h [hunk 35]
```diff
@@ -0,0 +1,57 @@
...
+    void SetupKeyMaps();
+
+    /// Sets analog deadzones
+    void SetDeadZone(int range);
+
+    /// Checks if analog input is within the dead zone
+    bool CheckDeadZone(int range) const;
+
+private:
+    SDL_Joystick* jpad;
```
Comment (neobrain) [inline src/input_common/sdl_input/sdl_input.h]: unique_ptr

File: src/input_common/sdl_input/sdl_input.h [hunk 36]
```diff
@@ -0,0 +1,57 @@
...
+#include <SDL2/SDL.h>
+#include <iostream>
+#include "input_common/input_common.h"
+
+namespace InputCommon {
+
+class SDLController final : public ControllerBase {
+public:
+    SDLController();
+    ~SDLController();
```
Comment (neobrain) [inline src/input_common/sdl_input/sdl_input.h]: shouldn't this be marked override, too?

File: src/input_common/sdl_input/sdl_input.h [hunk 37]
```diff
@@ -0,0 +1,57 @@
...
+    bool CheckDeadZone(int range) const;
+
+private:
+    SDL_Joystick* jpad;
+
+    /// SDL event used for polling
+    SDL_Event input_event;
+
+    /// Index of joystick, used for opening/closing a joystick
+    int index;
```
Comment (neobrain) [inline src/input_common/sdl_input/sdl_input.h]: Use unsigned for values which aren't ever intended to have a sign.

File: src/input_common/sdl_input/sdl_input.h [hunk 38]
```diff
@@ -0,0 +1,57 @@
...
+    void Shutdown() override;
+
+    /// Processes input when polling
+    void ProcessInput();
+
+    /// Sets up keymaps from input configuration
+    void SetupKeyMaps();
+
+    /// Sets analog deadzones
+    void SetDeadZone(int range);
```
Comment (neobrain) [inline src/input_common/sdl_input/sdl_input.h]: Needs documentation on the "range" parameter.

File: src/input_common/input_common.h [hunk 39]
```diff
@@ -0,0 +1,60 @@
...
+
+    /// Returns internal pad state
+    Service::HID::PadState GetPadState() const;
+
+    bool activated;
```
Comment (neobrain) [inline src/input_common/input_common.h]: public members *need* to be documented properly.

File: src/input_common/sdl_input/sdl_input.cpp (hunk 32 above)
Comment (neobrain) [inline src/input_common/sdl_input/sdl_input.cpp]: Agreeing that configurability should be taken care of in the initial PR rather than in a followup one.

File: src/input_common/sdl_input/sdl_input.cpp [hunk 40]
```diff
@@ -0,0 +1,238 @@
...
+    KeyMap::SetKeyMapping({100, joypad_id}, Service::HID::PAD_A);
+    KeyMap::SetKeyMapping({101, joypad_id}, Service::HID::PAD_B);
+    KeyMap::SetKeyMapping({102, joypad_id}, Service::HID::PAD_X);
+    KeyMap::SetKeyMapping({103, joypad_id}, Service::HID::PAD_Y);
+    KeyMap::SetKeyMapping({104, joypad_id}, Service::HID::PAD_R);
+    KeyMap::SetKeyMapping({105, joypad_id}, Service::HID::PAD_L);
+    KeyMap::SetKeyMapping({0, joypad_id}, Service::HID::PAD_ZL);
+    KeyMap::SetKeyMapping({0, joypad_id}, Service::HID::PAD_ZR);
+
+    KeyMap::SetKeyMapping({107, joypad_id}, Service::HID::PAD_START);
```
Comment (neobrain) [inline src/input_common/sdl_input/sdl_input.cpp]: What's with all these hardcoded numbers?

File: src/input_common/sdl_input/sdl_input.cpp [hunk 41]
```diff
@@ -0,0 +1,238 @@
...
+}
+
+SDLController::~SDLController() {
+    Shutdown();
+}
+
+bool SDLController::Init() {
+
+    //Attempt to initialize SDL with joystick only
+    if (SDL_Init(SDL_INIT_JOYSTICK) != 0) {
```
Comment (neobrain) [inline src/input_common/sdl_input/sdl_input.cpp]: Wow, I didn't know SDL actually still calls this stuff joystick, too...

File: src/input_common/input_common.cpp [hunk 42]
```diff
@@ -0,0 +1,85 @@
...
+
+    //Reschedule recurrent event
+    CoreTiming::ScheduleEvent(frame_ticks - cycles_late, input_event);
+}
+
+std::string ControllerBase::GetDeviceName() const {
+    return device_name;
+}
+
+Service::HID::PadState ControllerBase::GetPadState() const {
```
Comment (neobrain) [inline src/input_common/input_common.cpp]: should return a const reference, probably.

File: src/core/system.cpp (hunk 10 above)
Comment (neobrain) [inline src/core/system.cpp]: Will this code actually work if SDL2 is not installed on the system?

File: src/core/hle/service/hid/hid.cpp (hunk 9 above)
Comment (neobrain) [inline src/core/hle/service/hid/hid.cpp]: This looks really fishy, and by design prevents combining e.g. gamepad button input and mouse touchscreen input until InputCommon is in some way merged with EmuWindow. Note that I don't think with the current design it is sensible to move keyboard+mouse input from EmuWindow to InputCommon (I'll probably write a comment on that later on).

File: src/input_common/input_common.h [hunk 43]
```diff
@@ -0,0 +1,60 @@
...
+
+protected:
+    /// Internal name of currently selected device
+    std::string device_name;
+
+    /// Internal view this specific object will have of 3DS input states
+    ControllerState controller;
```
Comment (neobrain) [inline src/input_common/input_common.h]: Why are the touch screen and circlepad inputs not exposed publically in this interface?

File: src/input_common/sdl_input/sdl_input.cpp [hunk 44]
```diff
@@ -0,0 +1,238 @@
...
+void SDLController::ProcessInput() {
+    //Use a unique id for joystick input
+    int pad_id = 0;
+    int val = 0;
+
+    //Generate the id based on input from an event
+    switch (input_event.type) {
+        //Joystick button press events
+        case SDL_JOYBUTTONDOWN:
+            pad_id = 100;
```
Comment (neobrain) [inline src/input_common/sdl_input/sdl_input.cpp]: Where do these constants come from?

File: src/input_common/sdl_input/sdl_input.cpp [hunk 45]
```diff
@@ -0,0 +1,238 @@
...
+
+            break;
+
+        //Joystick hat motion events
+        case SDL_JOYHATMOTION:
+            pad_id = 300;
+            pad_id += input_event.jhat.hat * 4;
+
+            switch (input_event.jhat.value) {
+                case SDL_HAT_CENTERED:
```
Comment (neobrain) [inline src/input_common/sdl_input/sdl_input.cpp]: This whole switch case looks like voodoo. What's going on there?

**Comment by neobrain** [comment]
Made an initial pass over the code. Looking mostly good modulo coding style issues ;)

There's one design change I would like to see implemented: Input grabbing shouldn't be core emulator work but instead frontend work, which is one of the reasons it's part of EmuWindow to begin with. Note that keyboard and mouse handling should in any case stay in EmuWindow since they are handled by the GUI toolkit used by the frontend (not getting those inputs from the GUI is a mess). Moving things to EmuWindow is hence consistent and beneficial overall, since it gets rid of the ugly `g_user_input ? InputCommon::... : EmuWindow:... ` checks in core code. InputCommon interface itself need not be changed (much) for this, but the place where Poll() gets called should be moved to a thread managed by the EmuWindow implementations and the PadState/ControllerState should be publicly exposed only by EmuWindow.

File: src/input_common/sdl_input/sdl_input.cpp (hunk 45 above)
Comment (shonumi) [inline src/input_common/sdl_input/sdl_input.cpp]: I guess I should better commented about pad_ids. They're a way I came up with for SDL to assign unique numbers  for controller input. It was designed to be simple and easy to remember (unlike using horrible hashes like Mednafen iirc). 

Essentially the first analog input is assigned values 200, 201, 202, 203, corresponding for left, right, up, and down movements. Whenever the user moves the stick to the left, it's physically impossible to activate any input that depends on right movement. With pad_ids, we can always tell which id is "opposite" by looking at whether the id is odd or even (modulo 2). If it's even, we need to release the next highest pad_id (200 -> 201); if it's odd we need to release the next the next lowest pad_id (201 -> 200)

So better commenting? Hope that clears up the "voodoo". I was going to make a gist properly detailing how pads worked, since I didn't feel it quite right to make it a wiki item on citra (but I could have made it a wiki page on the branch >_<)

File: src/input_common/sdl_input/sdl_input.cpp (hunk 44 above)
Comment (shonumi) [inline src/input_common/sdl_input/sdl_input.cpp]: They come from SDL.h (which pulls SDL_joystick.h)

File: src/input_common/sdl_input/sdl_input.cpp (hunk 40 above)
Comment (shonumi) [inline src/input_common/sdl_input/sdl_input.cpp]: Initial work until they can be read from a file. Does Citra have some generic file parsing set up already? I would prefer not to have to come up with a custom solution.

File: src/input_common/sdl_input/sdl_input.cpp (hunk 32 above)
Comment (shonumi) [inline src/input_common/sdl_input/sdl_input.cpp]: Yeah, you're right. Looks like I need to switch to the SDL GameController API.

File: externals/cmake-modules/FindSDL2.cmake (hunk 29 above)
Comment (shonumi) [inline externals/cmake-modules/FindSDL2.cmake]: Fwiw, this was a copy+paste job from Dolphin's CMake module :p I'll change it soon in this PR.

File: src/input_common/input_common.cpp (hunk 12 above)
Comment (shonumi) [inline src/input_common/input_common.cpp]: I'm fine with making SDL2 mandatory. I was just concerned that we might want to use other APIs in the future. Is this a valid concern? I remember MaJoR explaining why SDL2 was removed as an input backend for Dolphin on Windows (something about webcams getting detected as controllers) but I'm sure we can avoid or workaround any pitfalls.

File: src/core/system.cpp (hunk 10 above)
Comment (shonumi) [inline src/core/system.cpp]: @bunnei - Agreed. Sorry, as you can see this PR is more a Preliminary Review rather than a Pull Request :/ It's rough, I apologize for stuff like this.

@neobrain - It compiles just fine with Travis CI. The CMake log shows it can't find SDL but builds properly.

File: src/input_common/input_common.h (hunk 43 above)
Comment (shonumi) [inline src/input_common/input_common.h]: Honestly overlooked that :/ Wow

File: src/core/hle/service/hid/hid.cpp (hunk 9 above)
Comment (shonumi) [inline src/core/hle/service/hid/hid.cpp]: @neobrain - Actually touchscreen input + gamepad input works. Touchscreen input (from EmuWindow) does not come from it's PadState. That's handled by separate members. I tested this from my end on a few games. It does prevent keyboard input from EmuWindow and gamepad input from occurring at the same time. But yes, the code here is an ugly temp solution...

**Comment by shonumi** [comment]
@neobrain - I kind of struggled finding a place for Poll() initially :p I'm still not sure where everyone stands on EmuWindow though. Some don't think it should handle input (at all), so I'd like to have a definitive direction in regards to that. 

I'm totally all for Polling in EmuWindow, but it rubs me the wrong way having to directly process "raw" input in EmuWindow ([like this](https://github.com/citra-emu/citra/blob/master/src/common/emu_window.cpp#L9)). At least, processing directly seems weird to do when that's InputCommon's job. It would be necessary to write out code for each specific GUI we have, but if bunnei says we can eliminate GLFW, that means just Qt for the time being.

So in short, how involved should EmuWindow be with input (Polling, Pass "raw" input vs. processing it directly, none at all)?

File: src/input_common/sdl_input/sdl_input.cpp (hunk 13 above)
Comment (shonumi) [inline src/input_common/sdl_input/sdl_input.cpp]: If it were passed it as an argument to SDLController::Poll(), what would be the best way to call it from [here](https://github.com/shonumi/citra/blob/input_common/src/input_common/input_common.cpp#L30)?

File: src/input_common/sdl_input/sdl_input.cpp (hunk 22 above)
Comment (shonumi) [inline src/input_common/sdl_input/sdl_input.cpp]: About the else parts, like [this](https://gist.github.com/shonumi/a4725247ebaa3c6d9a7b)?

File: src/input_common/input_common.cpp (hunk 21 above)
Comment (shonumi) [inline src/input_common/input_common.cpp]: How about using unique pointer (as neobrain suggested) and just doing a simple check like `if (g_user_input)`, or would you still prefer my making an `IsValid()` function?

File: src/input_common/sdl_input/sdl_input.h (hunk 35 above)
Comment (shonumi) [inline src/input_common/sdl_input/sdl_input.h]: When using unique_ptr, how does one set the stored pointer when that pointer comes from a function. See [here](https://github.com/shonumi/citra/blob/input_common/src/input_common/sdl_input/sdl_input.cpp#L37), I'm having trouble figuring out how to do it properly.

Comment (Subv) [inline src/input_common/sdl_input/sdl_input.h]: You cant use unique_ptr here, this code is not the owner of the pointer, and it should not be released by calling delete on it (You should use SDL_JoystickClose on it, and you're already doing that as far as i can see).

**Comment by neobrain** [comment]
@shonumi My opinion is that all of it should be done in EmuWindow one way or the other. You're free to just have the EmuWindow implementation refer to InputCommon however. That way we can have a single InputCommon framework for all EmuWindow implementations.

**Comment by neobrain** [comment]
Unfortunately, it seems like @shonumi will not be able to work on this any further. (correct me if I'm wrong!)

**Comment by shonumi** [comment]
Sorry, I'm still dealing with personal matters atm; it leaves very little time for other projects. Hopefully that will change in the future.
